src/main/java/com/github/rschmitt/dynamicobject/internal/Instances.java

```diff
@@ -29,7 +29,15 @@
     private static <D extends DynamicObject<D>> D createIndyProxy(Map map, Class<D> type) {
         ensureInitialized(type);
         try {
-            Object proxy = proxyCache.computeIfAbsent(type, Instances::createProxy)
+            // Not using ConcurrentHashMap#computeIfAbsent to avoid locking: JDK-8161372
+            DynamicProxy dynamicProxy;
+            if (proxyCache.containsKey(type)) {
+                dynamicProxy = proxyCache.get(type);
+            } else {
+                dynamicProxy = createProxy(type);
```

Comment:
It appears to me that this can result in multiple `DynamicProxy` instances being created for the same `type`. This is _probably_ harmless, but why not prevent it? The `else` branch can use pessimistic locking, synchronizing on `type` instead of invoking `#computeIfAbsent`:

```java
synchronized (type) {
    dynamicProxy = proxyCache.get(type);
    if (dynamicProxy == null) {
        dynamicProxy = createProxy(type);
        proxyCache.put(type, dynamicProxy);
    }
}
```

Reply:
I believe that would still end up locking on the same class object if deserializing the same type of object.

I could instead do something like this:

```
            if (!proxyCache.containsKey(type)) {
                proxyCache.putIfAbsent(type, createProxy(type));
            }
            Object proxy = proxyCache.get(type)
                .constructor()
                //...
```

This can create multiple at the start, but only the first one will ever succeed in being used. Initial calls would lock due to putIfAbsent, while subsequent calls won't as they will see that the entry is already in the cache then do a get.

Reply:
> I believe that would still end up locking on the same class object if deserializing the same type of object.

In the initialization path, yes, that's the intention. What's the objection?

Reply:
Removing computeIfAbsent is meant to avoid locking when initializing the same type of object in parallel. Removing the synchronized keyword from `ensureInitialized` is meant to avoid locking when initializing objects of any type.

If we synchronize on type, it would be locking just as computeIfAbsent did.

Reply:
But the difference is that synchronization was being encountered *on every call* (first during the call to `ensureInitialized`, then during the call to `computeIfAbsent`), whereas now it's only being encountered during initialization. Keep in mind that the code snippet I suggested was only for the `else` branch, which means that in context the code would look like this:

```java
        ensureInitialized(type);
        try {
            if (proxyCache.containsKey(type)) {
                dynamicProxy = proxyCache.get(type);
            } else {
                synchronized (type) {
                    dynamicProxy = proxyCache.get(type);
                    if (dynamicProxy == null) {
                        dynamicProxy = createProxy(type);
                        proxyCache.put(type, dynamicProxy);
                    }
                }
            }
            // and so on...
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
```

This means that once the cache is warmed up, we only encounter lockless read operations; no synchronizations should occur.

Reply:
Oh I see. It was my mistake for overlooking that you meant this for the `else` block. That looks reasonable to me. I'll update this.